### PR TITLE
History type support

### PR DESCRIPTION
**Pull request by DJAndries**
Adds support for the new Chromium history type.

History sync entities will use the `ClientTagHash` as a primary key, instead of the `IdString`. Differing version numbers between old and new items will be ignored when updating.

### CODE REVIEW

File: schema/dynamodb/table.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
   ],
   "GlobalSecondaryIndexes": [
     {
-      "IndexName": "ClientIDDataTypeMtimeIndex",
+      "IndexName": "ClientIDDataTypeMtimeIndexV2",
```
Comment (DJAndries) [inline schema/dynamodb/table.json:15]: a couple db changes are needed for this to work:

- the creation of this new index (and subsequent deletion of the old index)
- enabling ttl for the `ExpirationTime` attribute

File: command/command.go [hunk 2]
```diff
@@ -241,8 +242,25 @@ func handleCommitRequest(cache *cache.Cache, commitMsg *sync_pb.CommitMessage, c
 		}
 
 		oldVersion := *entityToCommit.Version
+		isUpdateOp := oldVersion != 0
 		*entityToCommit.Version = *entityToCommit.Mtime
-		if oldVersion == 0 { // Create
+		if *entityToCommit.DataType == historyTypeID {
+			// Check if item exists using client_unique_tag
+			isUpdateOp, err = db.HasItem(clientID, *entityToCommit.ClientDefinedUniqueTag)
+			log.Info().
```
Comment (yrliou) [inline command/command.go:250]: It was for debugging, please help remove this.

File: command/command.go [hunk 3]
```diff
@@ -241,8 +242,25 @@ func handleCommitRequest(cache *cache.Cache, commitMsg *sync_pb.CommitMessage, c
...
+				Str("Tag", *entityToCommit.ClientDefinedUniqueTag).
+				Bool("IsUpdate", isUpdateOp).
+				Msg("History Info")
+			if err != nil {
+				log.Error().Err(err).Msg("Insert sync entity failed")
+				rspType := sync_pb.CommitResponse_TRANSIENT_ERROR
+				entryRsp.ResponseType = &rspType
+				entryRsp.ErrorMessage = aws.String(fmt.Sprintf("Insert sync entity failed: %v", err.Error()))
```
Comment (yrliou) [inline command/command.go:258]: Please help update the message to match the case here.

File: datastore/sync_entity.go [hunk 4]
```diff
@@ -159,7 +165,7 @@ func (dynamo *Dynamo) InsertSyncEntity(entity *SyncEntity) (bool, error) {
 		return false, fmt.Errorf("error building expression to insert sync entity: %w", err)
 	}
 
-	if entity.ClientDefinedUniqueTag != nil {
+	if entity.ClientDefinedUniqueTag != nil && *entity.DataType != historyTypeID {
```
Comment (yrliou) [inline datastore/sync_entity.go:168]: Please update the comment above this function to mention the history type case does not use tag item.

File: datastore/sync_entity.go [hunk 5]
```diff
@@ -253,6 +259,28 @@ func (dynamo *Dynamo) HasServerDefinedUniqueTag(clientID string, tag string) (bo
 	return out.Item != nil, nil
 }
 
+func (dynamo *Dynamo) HasItem(clientID string, ID string) (bool, error) {
+	primaryKey := PrimaryKey{ClientID: clientID, ID: ID}
+	key, err := dynamodbattribute.MarshalMap(primaryKey)
+
+	if err != nil {
+		return false, fmt.Errorf("error marshalling key to check if client tag existed: %w", err)
```
Comment (yrliou) [inline datastore/sync_entity.go:267]: s/client tag/item

File: datastore/sync_entity.go [hunk 6]
```diff
@@ -253,6 +259,28 @@ func (dynamo *Dynamo) HasServerDefinedUniqueTag(clientID string, tag string) (bo
...
+
+	input := &dynamodb.GetItemInput{
+		Key:                  key,
+		ProjectionExpression: aws.String(projPk),
+		TableName:            aws.String(Table),
+	}
+
+	out, err := dynamo.GetItem(input)
+	if err != nil {
+		return false, fmt.Errorf("error calling GetItem to check if client tag existed: %w", err)
```
Comment (yrliou) [inline datastore/sync_entity.go:278]: s/client tag/item

File: datastore/sync_entity.go [hunk 7]
```diff
@@ -745,16 +779,27 @@ func CreateDBSyncEntity(entity *sync_pb.SyncEntity, cacheGUID *string, clientID
 		originatorClientItemID = entity.IdString
 	}
 
-	now := aws.Int64(utils.UnixMilli(time.Now()))
+	if dataType == historyTypeID {
```
Comment (yrliou) [inline datastore/sync_entity.go:782]: Please add a comment above this line for what we're doing for history and delete directive here.

**Review by yrliou — COMMENTED** [review]
Overall looks good, just some minor comments for readability.
Also please make sure to have a green CI test.

File: command/command.go [hunk 8]
```diff
@@ -241,8 +242,21 @@ func handleCommitRequest(cache *cache.Cache, commitMsg *sync_pb.CommitMessage, c
...
-		if oldVersion == 0 { // Create
+		if *entityToCommit.DataType == historyTypeID {
+			// Check if item exists using client_unique_tag
+			isUpdateOp, err = db.HasItem(clientID, *entityToCommit.ClientDefinedUniqueTag)
+			if err != nil {
+				log.Error().Err(err).Msg("Insert history sync entity failed")
+				rspType := sync_pb.CommitResponse_TRANSIENT_ERROR
+				entryRsp.ResponseType = &rspType
+				entryRsp.ErrorMessage = aws.String(fmt.Sprintf("Insert history sync entity failed: %v", err.Error()))
+				continue
```
Comment (yrliou) [inline command/command.go:255]: nit: Maybe `Error checking if client_unique_tag exists` rather than `Insert history sync entity failed` is more clear?

**Review by yrliou — APPROVED** [review]

**Pull request merged**